### PR TITLE
[Buttons] Implement a color themer for an MDCRaisedButton

### PR DESCRIPTION
**Pull request by yarneo**
Addition of a unit test, and updated example to use the new themer.

Pivotal: https://www.pivotaltracker.com/story/show/156640797

simulator screen shot - iphone 8 - 2018-04-10 at 16 13 40

### CODE REVIEW

File: components/Buttons/src/ColorThemer/MDCButtonColorThemer.h [hunk 1]
```diff
@@ -30,6 +30,14 @@
 + (void)applySemanticColorScheme:(nonnull id<MDCColorScheming>)colorScheme
                         toButton:(nonnull MDCButton *)button;
 
+/**
+ Applies a color scheme to theme to a MDCRaisedButton.
```
Comment (romoore) [inline components/Buttons/src/ColorThemer/MDCButtonColorThemer.h]: Super-nit: Should be "...an MDCRaisedButton."

File: components/Buttons/src/ColorThemer/MDCButtonColorThemer.h [hunk 2]
```diff
@@ -30,6 +30,14 @@
...
+
+ @param colorScheme The color scheme to apply to MDCRaisedButton.
```
Comment (romoore) [inline components/Buttons/src/ColorThemer/MDCButtonColorThemer.h]: LINT: Missing `@param` for `raisedButton`
Nit: This comment should be, "...apply to @c raisedButton."

File: components/Buttons/src/ColorThemer/MDCButtonColorThemer.m [hunk 3]
```diff
@@ -33,6 +33,11 @@ + (void)applySemanticColorScheme:(nonnull id<MDCColorScheming>)colorScheme
   button.disabledAlpha = 1.f;
 }
 
++ (void)applySemanticColorScheme:(nonnull id<MDCColorScheming>)colorScheme
```
Comment (romoore) [inline components/Buttons/src/ColorThemer/MDCButtonColorThemer.m:33]: Why do we have a specific method for `toRaisedButton:` if there is no such class and it just calls the general themer?

File: components/Buttons/tests/unit/ButtonsColorThemerTests.m [hunk 4]
```diff
@@ -62,4 +62,41 @@ - (void)testMDCButtonColorThemer {
   XCTAssertEqual(button.disabledAlpha, 1.f);
 }
 
+- (void)testMDCRaisedButtonColorThemer {
```
Comment (romoore) [inline components/Buttons/tests/unit/ButtonsColorThemerTests.m]: Nit: The name should probably be something like `-testApplySemanticColorSchemeToRaisedButton`. There is no `MDCRaisedButtonColorThemer` class.

File: components/Buttons/tests/unit/ButtonsColorThemerTests.m [hunk 5]
```diff
@@ -62,4 +62,41 @@ - (void)testMDCButtonColorThemer {
...
+  colorScheme.onSurfaceColor = UIColor.blueColor;
+  [button setTitleColor:UIColor.whiteColor forState:UIControlStateNormal];
+  [button setTitleColor:UIColor.greenColor forState:UIControlStateHighlighted];
+  [button setTitleColor:UIColor.blueColor forState:UIControlStateSelected];
+  [button setTitleColor:UIColor.grayColor forState:UIControlStateDisabled];
+  [button setBackgroundColor:UIColor.purpleColor forState:UIControlStateNormal];
+  [button setBackgroundColor:UIColor.redColor forState:UIControlStateHighlighted];
+  [button setBackgroundColor:UIColor.blueColor forState:UIControlStateSelected];
+  [button setBackgroundColor:UIColor.darkGrayColor forState:UIControlStateDisabled];
+
```
Comment (romoore) [inline components/Buttons/tests/unit/ButtonsColorThemerTests.m]: Please add a few more states and verify that they remain unaffected by the tests.  For example, `UIControlStateSelected | UIControlStateHighlighted`

File: components/Buttons/tests/unit/ButtonsColorThemerTests.m [hunk 6]
```diff
@@ -62,4 +62,41 @@ - (void)testMDCButtonColorThemer {
...
+  // Where
+  [MDCButtonColorThemer applySemanticColorScheme:colorScheme toRaisedButton:button];
+
+  // Then
+  XCTAssertEqual([button titleColorForState:UIControlStateNormal], colorScheme.onPrimaryColor);
+  XCTAssertEqual([button titleColorForState:UIControlStateHighlighted], colorScheme.onPrimaryColor);
+  XCTAssertEqual([button titleColorForState:UIControlStateSelected], colorScheme.onPrimaryColor);
+  XCTAssert(
```
Comment (romoore) [inline components/Buttons/tests/unit/ButtonsColorThemerTests.m]: I believe XCTAssertEqualObjects works with colors just fine as long as the alpha is the same precision (single/double) in both initializer invocations.

File: components/Buttons/src/ColorThemer/MDCButtonColorThemer.m [hunk 7]
```diff
@@ -33,6 +33,11 @@ + (void)applySemanticColorScheme:(nonnull id<MDCColorScheming>)colorScheme
...
+                        toRaisedButton:(nonnull MDCRaisedButton *)raisedButton {
```
Comment (jverkoey) [inline components/Buttons/src/ColorThemer/MDCButtonColorThemer.m]: minor nit: indentation

**Review by romoore — CHANGES_REQUESTED** [review]

File: components/Buttons/tests/unit/ButtonsColorThemerTests.m [hunk 8]
```diff
@@ -62,4 +62,41 @@ - (void)testMDCButtonColorThemer {
...
+      CGColorEqualToColor([button titleColorForState:UIControlStateDisabled].CGColor,
+                          [colorScheme.onSurfaceColor colorWithAlphaComponent:0.26f].CGColor));
+  XCTAssertEqual([button backgroundColorForState:UIControlStateNormal], colorScheme.primaryColor);
+  XCTAssertEqual([button backgroundColorForState:UIControlStateHighlighted],
+                 colorScheme.primaryColor);
+  XCTAssertEqual([button backgroundColorForState:UIControlStateSelected], colorScheme.primaryColor);
+  XCTAssert(
+      CGColorEqualToColor([button backgroundColorForState:UIControlStateDisabled].CGColor,
+                          [colorScheme.onSurfaceColor colorWithAlphaComponent:0.12f].CGColor));
+  XCTAssertEqual(button.disabledAlpha, 1.f);
```
Comment (romoore) [inline components/Buttons/tests/unit/ButtonsColorThemerTests.m]: Please use XCTAssertEqualWithAccuracy for floating point comparison.

**Comment by ianegordon** [comment]
Please rebase / fix merge conflicts when addressing @romoore comments.

**Comment by yarneo** [comment]
some rebase issues, will reopen once ready.

**Pull request closed without merging**